_posts/2019-08-12-https-nginx.md

```diff
@@ -0,0 +1,209 @@
+```
+RUN apt-get install -y certbot python-certbot-nginx 
+```
+
+Then, we rebuild the image and run a container from it. At this point we installed the certbot agent. We now run the command in the container spawned from this image:
+```bash
+  $ docker ps
+  $ docker exec -it <NGINX_CONTAINER_ID> bash
+  $ letsencrypt certonly -a webroot --webroot-path=/var/www/folder_where_challenge_will_be_put -d www.domain.co.uk
+```
+
+If we run this command, the challenge will fail. As we said previously, the challenge consists in retrieving a file that is created on our machine. We need to change the nginx config to serve the folder where the file has been created, in our example it's `/var/www/folder_where_challenge_will_be_put/`. 
+We need to add this part to the config file:
+
+```
+  location ~ /.well-known {
+    allow all;
+    root /var/www/folder_where_challenge_will_be_put/;
+  }
+```
+
+If the last command succeeds, we generated the certificates which have been installed at `/etc/letsencrypt/live/www.domain.co.uk/`.
+
+You can find the code of this step [here](https://github.com/nabilbouab/tech-blog-labs/tree/step-2)
+
+
+## Step 3: Make HTTPS work
+
+This last step will consist in two sub steps:
+  * We first need to make a redirection from HTTP to HTTPS
+  * We need to make NGINX do the encryption/decryption by providing it access to the generated certificates
+
+![SSL termination diagram](/assets/nginx-ssl-termination.png)
+
+In order to do the ssl termination, NGINX needs to be able to access the certificates which will be used to encrypt and decrypt the data. We need to change the NGINX config to point it to the newly generated certificates and make the redirection that way:
+
+```
+worker_processes  1;
+
+events {
+	worker_connections  1024;
+}
+
+
+http {
+	include       mime.types;
+	default_type  application/octet-stream;
+
+	sendfile        on;
+
+	keepalive_timeout  65;
+
+
+	upstream public {
+		server web:3000;
+	}
+
+	server {
+		listen 80;
+		server_name our-awesome-domain.fr;
+
+		location ~ /.well-known {
+			allow all;
+			root /var/www/our-awesome-domain.fr/;
+		}
+
+    location /favicon.ico {
+      root /var/www/;
+    }
+
+    location / {
+      return 301 https://$host$request_uri;
+    }
+	}
+
+  server {
+    listen 443 ssl;
+    server_name our-awesome-domain.fr;
+
+    ssl_certificate /etc/letsencrypt/live/www.our-awesome-domain.fr/fullchain.pem;
+    ssl_certificate_key /etc/letsencrypt/live/www.our-awesome-domain.fr/privkey.pem;
+
+    ssl_session_cache shared:SSL:50m;
+    ssl_session_timeout 10m;
+
+    # ssl_dhparam /etc/ssl/dhparams/dhparam.pem;
+
+    ssl_protocols TLSv1.1 TLSv1.2;
+    ssl_ciphers ECDH+AESGCM:ECDH+AES256:ECDH+AES128:DES-CBC3-SHA:!ADH:!AECDH:!MD5;
+    ssl_prefer_server_ciphers on;
+
+    resolver 8.8.8.8 8.8.4.4;
+    ssl_stapling on;
+    ssl_stapling_verify on;
+
+    add_header Strict-Transport-Security "max-age=31536000; includeSubdomains; preload";
```

Comment:
In the same vein as my comment on DH params: I know that people will copy paste that and wonder why they have issues.
This enables HSTS, I am a big fan of HSTS. What this header does, is to tell the clients to default to HTTPS on this website for the next year (31536000 seconds). What that means is that the website becomes completely unusable if you have certificate issues and you have to do HTTP for some time (because now all clients will refuse to do HTTP!). More: if parts of the website's assets are loaded over HTTP (because legacy) they will simply stop loading. 
I would leave this line here (because HSTS is amazing) but I would put a comment saying to only enable this if the HTTPS setup is stable.